File: objectModel/TypeScript/Persistence/CdmFolder/EntityPersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { CdmFolder } from '..';
import {
    CdmAttributeItem,
    CdmCorpusContext,
    CdmEntityDefinition,
    cdmObjectType,
    CdmTraitGroupReference,
    CdmTraitReference,
    CdmTraitReferenceBase,
    copyOptions,
    resolveOptions
} from '../../internal';
import * as copyDataUtils from '../../Utilities/CopyDataUtils';
import {
    AttributeContext,
    AttributeGroupReference,
    AttributeResolutionGuidance,
    Entity,
    EntityAttribute,
    EntityReferenceDefinition,
    TraitGroupReference,
    TraitReference,
    TypeAttribute
} from './types';
import * as utils from './utils';

export class EntityPersistence {
    public static fromData(ctx: CdmCorpusContext, object: Entity): CdmEntityDefinition {
        if (!object) {
            return undefined;
        }

        const entity: CdmEntityDefinition = ctx.corpus.MakeObject(cdmObjectType.entityDef, object.entityName);
        entity.extendsEntity = CdmFolder.EntityReferencePersistence.fromData(ctx, object.extendsEntity);
        entity.extendsEntityResolutionGuidance =
            CdmFolder.AttributeResolutionGuidancePersistence.fromData(ctx, object.extendsEntityResolutionGuidance);

        entity.explanation = utils.propertyFromDataToString(object.explanation);

        utils.addArrayToCdmCollection<CdmTraitReferenceBase>(
            entity.exhibitsTraits,
            utils.createTraitReferenceArray(ctx, object.exhibitsTraits)
        );

        entity.sourceName = utils.propertyFromDataToString(object.sourceName);
        entity.displayName = utils.propertyFromDataToString(object.displayName);
        entity.description = utils.propertyFromDataToString(object.description);
        entity.version = utils.propertyFromDataToString(object.version);
        entity.cdmSchemas = object.cdmSchemas;

        if (object.attributeContext) {
            entity.attributeContext = CdmFolder.AttributeContextPersistence.fromData(ctx, object.attributeContext);
        }

        utils.addArrayToCdmCollection<CdmAttributeItem>(entity.attributes, utils.createAttributeArray(ctx, object.hasAttributes, entity.entityName));

        return entity;
    }

    public static toData(instance: CdmEntityDefinition, resOpt: resolveOptions, options: copyOptions): Entity {
        const exhibitsTraits: CdmTraitReferenceBase[] = instance.exhibitsTraits ?
            instance.exhibitsTraits.allItems.filter(
                (trait: CdmTraitReferenceBase) => trait instanceof CdmTraitGroupReference || !(trait as CdmTraitReference).isFromProperty) : undefined;
        const object: Entity = {
            explanation: instance.explanation,
            entityName: instance.entityName,
            extendsEntity: instance.extendsEntity ?
                instance.extendsEntity.copyData(resOpt, options) as (string | EntityReferenceDefinition) : undefined,
            extendsEntityResolutionGuidance: instance.extendsEntityResolutionGuidance ?
                instance.extendsEntityResolutionGuidance.copyData(resOpt, options) as AttributeResolutionGuidance : undefined,
            exhibitsTraits: copyDataUtils.arrayCopyData<string | TraitReference | TraitGroupReference>(resOpt, exhibitsTraits, options)
        };

        if (instance.sourceName) {
            object.sourceName = instance.getProperty('sourceName') as string;
        }
        if (instance.displayName) {
            object.displayName = instance.getProperty('displayName') as string;
        }
        if (instance.description) {
            object.description = instance.getProperty('description') as string;
        }
        if (instance.version) {
            object.version = instance.getProperty('version') as string;
        }
        if (instance.cdmSchemas) {
            object.cdmSchemas = instance.getProperty('cdmSchemas') as string[];
        }

        // after the properties so they show up first in doc
        if (instance.attributes) {
            object.hasAttributes =
            copyDataUtils.arrayCopyData<string | AttributeGroupReference | TypeAttribute | EntityAttribute>(
                    resOpt, instance.attributes, options);
            object.attributeContext = instance.attributeContext ?
                instance.attributeContext.copyData(resOpt, options) as AttributeContext : undefined;
        }

        return object;
    }
}

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/EntityPersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmEntityDefinition;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmPropertyName;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.AttributeContext;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.Entity;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.JMapper;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class EntityPersistence {

  public static CdmEntityDefinition fromData(final CdmCorpusContext ctx, final JsonNode obj) {
    if (obj == null) {
      return null;
    }

    final CdmEntityDefinition entity =
        ctx.getCorpus()
            .makeObject(
                CdmObjectType.EntityDef,
                obj.has("entityName")
                    ? obj.get("entityName").asText()
                    : null);

    entity.setExtendsEntity(EntityReferencePersistence.fromData(ctx, obj.get("extendsEntity")));

    entity.setExplanation(Utils.propertyFromDataToString(obj.get("explanation")));

    entity.setExtendsEntityResolutionGuidance(
        AttributeResolutionGuidancePersistence.fromData(
            ctx,
            obj.get("extendsEntityResolutionGuidance")));

    Utils.addListToCdmCollection(
        entity.getExhibitsTraits(),
        Utils.createTraitReferenceList(ctx, obj.get("exhibitsTraits")));
    if (obj.has("attributeContext")) {
      entity.setAttributeContext(
          AttributeContextPersistence.fromData(
              ctx,
              JMapper.MAP.convertValue(obj.get("attributeContext"),
                  AttributeContext.class)));
    }

    Utils.addListToCdmCollection(
        entity.getAttributes(),
        Utils.createAttributeList(ctx, obj.get("hasAttributes"), entity.getEntityName()));

    entity.setSourceName(Utils.propertyFromDataToString(obj.get("sourceName")));
    entity.setDisplayName(Utils.propertyFromDataToString(obj.get("displayName")));
    entity.setDescription(Utils.propertyFromDataToString(obj.get("description")));
    entity.setVersion(Utils.propertyFromDataToString(obj.get("version")));
    entity.setCdmSchemas(obj.has("cdmSchemas")
            ? null
            : JMapper.MAP.convertValue(obj.get("cdmSchemas"), new TypeReference<List<String>>() {
    }));

    return entity;
  }

  public static Entity toData(
      final CdmEntityDefinition instance,
      final ResolveOptions resOpt,
      final CopyOptions options) {
    final Entity obj = new Entity();
    obj.setExplanation(instance.getExplanation());
    obj.setEntityName(instance.getEntityName());
    obj.setExtendsEntity(Utils.jsonForm(instance.getExtendsEntity(), resOpt, options));
    obj.setExtendsEntityResolutionGuidance(
        Utils.jsonForm(
            instance.getExtendsEntityResolutionGuidance(),
            resOpt,
            options));
    obj.setExhibitsTraits(Utils.listCopyDataAsArrayNode(
        instance.getExhibitsTraits().getAllItems()
            .stream()
            .filter(trait -> !trait.isFromProperty())
            .collect(Collectors.toList()),
        resOpt,
        options));

    obj.setSourceName((String) instance.getProperty(CdmPropertyName.SOURCE_NAME));
    obj.setDisplayName((String) instance.getProperty(CdmPropertyName.DISPLAY_NAME));
    obj.setDescription((String) instance.getProperty(CdmPropertyName.DESCRIPTION));
    obj.setVersion((String) instance.getProperty(CdmPropertyName.VERSION));
    obj.setCdmSchemas((ArrayList<String>) instance.getProperty(CdmPropertyName.CDM_SCHEMAS));

    // after the properties so they show up first in doc
    obj.setAttributes(Utils.listCopyDataAsArrayNode(instance.getAttributes(), resOpt, options));
    obj.setAttributeContext(Utils.jsonForm(instance.getAttributeContext(), resOpt, options));

    return obj;
  }
}
